install.py

```diff
@@ -430,6 +430,19 @@ def build_legion(
             + setup_py_flags,
             cwd=legion_python_dir,
         )
+        # install the jupyter python files
```

Comment:
Does this process require ipython? If so, then this is implicitly adding ipython to our dependencies.

Reply:
We should try to avoid making ipython/jupyter a required dependency. I would even suggest splitting things off into a separate package if that cannot otherwise be avoided. Having separate packages just for jupyter kernels seems to be fairly common: https://pypi.org/search/?q=jupyter+kernel

Reply:
Another alternative to having a script to run that actually might be simpler from a packaging POV is to make this a module inside the package that can be called like:
```
python -m legate.jupyter install 
python -m legate.jupyter install <path to json>
```

We could do this with `legate/jupyter/__main__.py` module and then the ``legate/jupyter/` package dir is also a place to collect any jupyter-specific code now or in the future

Reply:
As a demonstration of the details:
```bash
❯ mkdir legate/jupyter

❯ touch legate/jupyter/__init__.py

❯ echo "print('legate jupyter main')" > legate/jupyter/__main__.py

❯ python -m legate.jupyter
legate jupyter main
```
@eddy16112 I'd be happy to have a call or otherwise talk over this idea, I think it would side-step package/install issues and leave a clear place for jupyter code to live

Reply:
This installation script is from Legion, do we want users to explicitly install the jupyter support of Legion?

Reply:
OK that's the important bit I somehow was missing. But I am not sure whether my suggestion has to change. I'm proposing that users are responsible for 
```
python -m legate.jupyter install <options>
```
from an *installed* (possibly editable) legate package.  Is there any reason why `legate/jupyter/__main__.py` could not do that work instead of putting it in `install.py` ?

I am assuming that legion's jupyter support installable as a package? Otherwise if the only way is to to invoke the `setup.py` in `join(legion_src_dir, "jupyter_notebook")` how is that supposed to work with conda package installs of legate?

Reply:
> I am assuming that legion's jupyter support installable as a package? Otherwise if the only way is to to invoke the setup.py in join(legion_src_dir, "jupyter_notebook") how is that supposed to work with conda package installs of legate?

Right, the setup.py from legion will just copy the `install_jupyter.py` and `legion_kernel_nocr.py` into the LEGATE_DIR/lib/python3.8/site_packages just like where other legion python files (legion_top.py and etc. ) are. 
Otherwise we will need to explicitly set the `os.path.append(legion_src_dir, "jupyter_notebook")`, which is what I did before.

Reply:
I still must be missing something. Let's say a user does:
```
conda install -c nvidia -c conda-forge -c legate cunumeric
```
They don't have a legion source checkout anywhere on their system. How does the jupyter kernel installation work for these users?

Reply:
> Are we going to be bundling the legion jupyter files in the legate conda pacage? 

Yes, we should do that. Once the legion jupyter files are copied into installation directory, a normal user can run 
```
LEGATE_DIR/bin/install_jupyter_kernel <--json=legate_jupyter.json>
```
to install the jupyter kernel

Reply:
That's why I have the code in the install.py to install the legion jupyter files into LEGETA_DIR
 ```       
         legion_jupyter_dir = os.path.join(legion_src_dir, "jupyter_notebook")
         verbose_check_call(
             [
                 sys.executable,
                 "setup.py",
                 "install",
                 "--prefix",
                 str(os.path.realpath(install_dir)),
             ]
             + setup_py_flags,
             cwd=legion_jupyter_dir,
         )
```

Reply:
OK I am still proposing that we put the code to do that in `legate/jupyter/__main__.py` then. That will afford the usage
```
python -m legate.jupyter install <options>
```
And that will work equally well whether:
* a conda package is installed
* a dev wheel is installed
* an editable dev wheel is install
* whether legate is installed in a regular site-package, or in some random `LEGATE_DIR`

I'd like to reiterate that after the cmake work is merged, a separate `LEGATE_DIR` will not be required and in fact would be highly discouraged. The default will be to install things in to the standard python site-packages. 

And if we really do want a "script" then that can be accomplished by defining an entry point for setuptools that points at `legate/jupyter/__main__.py`. That would add the script to the correct `bin` without lots of manual and fragile file copying on our part.

Reply:
I think we can do `python -m legate.jupyter`, but there are an issue that I do not know how to fix it. 

- If we are not in the `LEGATE_Source`, and `LEGATE_DIR` is not added into `PYTHONPATH`, then the `python -m legate.jupyter` won't work because python can not find the legate.jupyter. 
- If we are in the `LEGATE_Source`, and we will still ask users to set `LEGATE_DIR`, otherwise, python can not find the legion jupyter files pre-installed in the `LEGATE_DIR`

So the whole cmd will be:
```
LEGATE_DIR=<install_dir> python -m legate.jupyter <options>
```

Reply:
@eddy16112 I think once the cmake work is merged that it will not be necessary in the common case. In the common case, the packages are all installed in the *normal python environment site packages*, so everything  absolutely will be in the PYTHONPATH just like any normal Python package. This will be true for conda package installs, as well as default wheel and editable wheel dev installs. It will *not* be necessary for devs or users to set `LEGATE_DIR` -- that is one of the benefits of moving to a new system that uses regular python site-packages. 

The only case that will need worrying about then is if here is a dev install that set `--install-dir` explicitly. The option exists for backward compatibility with the current `install.py`, but should *not* be the usual workflow in the future. In that case, `python -m legate.jupyter` could also accept an optional `--install-dir` (or perhaps this metadata is available programmatically somehow cc @trxcllnt) 

I do still think it would prudent to sit on this PR for a little while longer, and update it after the cmake work goes in.

Reply:
OK, let's wait for the cmake. Currently, I have modified the installation into `python -m legate.jupyter`, but it still requires users to set LEGATE_DIR for now.